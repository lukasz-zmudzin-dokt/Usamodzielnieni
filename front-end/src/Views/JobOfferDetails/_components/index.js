import AddCvForm from './AddCvForm/AddCvForm';
import RemoveCv from './RemoveCv/RemoveCv';

export {
    AddCvForm,
    RemoveCv
};