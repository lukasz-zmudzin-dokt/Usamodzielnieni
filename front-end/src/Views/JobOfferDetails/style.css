.jobOfferDetails.container {
    margin-top: 100px;
    margin-bottom: 100px;
}
.jobOfferDetails h2.card-header {
    text-align: center;
}