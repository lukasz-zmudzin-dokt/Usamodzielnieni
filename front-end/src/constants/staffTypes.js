export const staffTypes = {
  VERIFICATION: "staff_verification",
  CV: "staff_cv",
  JOBS: "staff_jobs",
  BLOG_CREATOR: "staff_blog_creator",
  BLOG_MODERATOR: "staff_blog_moderator",
};
