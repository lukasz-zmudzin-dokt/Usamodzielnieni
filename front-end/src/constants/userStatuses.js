export const userStatuses = {
  VERIFIED: "Verified",
  REJECTED: "Not verified",
  AWAITING: "Waiting for verification",
};
