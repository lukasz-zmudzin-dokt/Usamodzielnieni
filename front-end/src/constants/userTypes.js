export const userTypes = {
  STANDARD: "standard",
  STAFF: "staff",
  EMPLOYER: "employer",
};
