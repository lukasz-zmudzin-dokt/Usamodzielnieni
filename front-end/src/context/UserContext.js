import React, { useState } from "react";
import Cookies from "universal-cookie";

export const UserContext = React.createContext({
  type: undefined, // Employer,Standard,Staff
  token: undefined,
  login: () => {},
  logout: () => {}
});
const cookies = new Cookies();

export const UserProvider = props => {
  const [type, setType] = useState(cookies.get("type"));
  const [token, setToken] = useState(cookies.get("token"));
  const user = {
    token,
    type,
    login: (newToken, newType) => {
      cookies.set("token", newToken, { path: "/" });
      cookies.set("type", newType, { path: "/" });
      setToken(newToken);
      setType(newType);
    },
    logout: () => {
      cookies.remove("token", { path: "/" });
      cookies.remove("type", { path: "/" });
      setToken(undefined);
      setType(undefined);
    }
  };
  return <UserContext.Provider value={user} {...props} />;
};
