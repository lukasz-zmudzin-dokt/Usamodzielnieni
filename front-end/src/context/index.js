export { UserContext, UserProvider } from "context/UserContext";
export { AlertContext, AlertProvider } from "context/AlertContext";
export { ChatContext, ChatProvider } from "context/ChatContext";
export {
  NotificationsContext,
  NotificationsProvider,
} from "context/NotificationsContext";
