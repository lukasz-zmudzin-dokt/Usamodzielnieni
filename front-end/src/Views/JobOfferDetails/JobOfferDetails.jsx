import React, { useState, useEffect, useContext } from 'react';
import { Container, Card, Alert, Row } from "react-bootstrap";
import { withRouter } from "react-router-dom";
import { UserContext } from "context";
import { DetailsItem } from 'components';
import { AddCvForm, RemoveOffer } from "./_components";
import { staffTypes } from "constants/staffTypes";
import proxy from "config/api";

const getOfferDetails = async (id, token) => {
  let url = `${proxy.job}job-offer/${id}`;
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json"
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status === 200) {
    return response.json().then(offer => mapOffer(offer));
  } else {
    throw response.status;
  }
}

const mapOffer = (offer) => ({
  id: offer.id,
  title: offer.offer_name,
  category: offer.category,
  type: offer.type,
  companyName: offer.company_name,
  companyAddress: `ul. ${offer.company_address.street} ${offer.company_address.street_number}, ${offer.company_address.postal_code} ${offer.company_address.city}`,
  voivodeship: offer.voivodeship,
  expirationDate: offer.expiration_date,
  description: offer.description
})


const JobOfferDetails = props => {
  const [offer, setOffer] = useState({});
  const [isOfferLoading, setIsOfferLoading] = useState(false);
  const [error, setError] = useState(false);
  const user = useContext(UserContext);

  useEffect(
    () => {
      const loadOffer = async (id, token) => {
        setIsOfferLoading(true);
        let loadedOffer;
        try {
          loadedOffer = await getOfferDetails(id, token);
        } catch (e) {
          console.log(e);
          loadedOffer = {};
          setError(true);
        }
        setOffer(loadedOffer);
        setIsOfferLoading(false);
      }
      loadOffer(props.match.params.id, user.token)
    },
    [props.match.params.id, user.token]
  );

  const msg = isOfferLoading ? <Alert variant="info">Ładowanie oferty...</Alert> :
              error && <Alert variant="danger">Wystąpił błąd podczas ładowania oferty.</Alert>
  console.log(offer);

  return (
    <Container className="jobOfferDetails">
      <Card>
      <Card.Header as="h2">Szczegóły oferty pracy</Card.Header>
      <Card.Body>
        { msg || (
          <div>
            <h3>{offer.title}</h3>
            <Row>
              <DetailsItem md="6" xl="4" label="Nazwa firmy">{offer.companyName}</DetailsItem>
              <DetailsItem md="6" xl="4" label="Adres firmy">{offer.companyAddress}</DetailsItem>
              <DetailsItem md="6" xl="4" label="Lokalizacja">{offer.voivodeship}</DetailsItem>
              <DetailsItem md="6" xl="4" label="Ważne do">{offer.expirationDate}</DetailsItem>
              <DetailsItem md="6" xl="4" label="Branża">{offer.category}</DetailsItem>
              <DetailsItem md="6" xl="4" label="Typ">{offer.type}</DetailsItem>
            </Row>
            <p>{offer.description}</p>
          </div>
        )}
        { user.type === 'Standard' && user.data?.status === 'Verified' && <AddCvForm id={props.match.params.id} user={user} /> }
        { user.type === 'Staff' && user.data?.group_type.includes(staffTypes.JOBS) && <RemoveOffer id={props.match.params.id} user={user} /> }
      </Card.Body>
      </Card>
    </Container>
  )
};

export default withRouter(JobOfferDetails);