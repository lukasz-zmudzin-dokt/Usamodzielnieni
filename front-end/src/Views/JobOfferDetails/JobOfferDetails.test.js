import React from "react";
import { render, waitForElement } from "@testing-library/react";
import JobOfferDetails from "./JobOfferDetails";
import { MemoryRouter } from "react-router-dom";
import { UserContext } from "context/UserContext";
import { userTypes } from "constants/userTypes";
import { userStatuses } from "constants/userStatuses";
import { staffTypes } from "constants/staffTypes";

jest.mock("./_components", () => ({
  AddCvForm: () => <div>AddCvForm</div>,
  RemoveOffer: () => <div>RemoveOffer</div>,
}));
jest.mock("constants/staffTypes", () => ({
  staffTypes: {
    JOBS: "jobs",
  },
}));

describe("JobOfferDetails", () => {
  let offer;
  let apiStatus;
  let match;
  let user;

  beforeAll(() => {
    match = { params: { id: "123" } };
    global.fetch = jest.fn().mockImplementation((input, init) => {
      return new Promise((resolve, reject) => {
        switch (init.method) {
          case "GET":
            resolve(
              apiStatus !== 200
                ? { status: apiStatus }
                : {
                    status: 200,
                    json: () => Promise.resolve(offer),
                  }
            );
            break;
          default:
            reject({});
            break;
        }
      });
    });
  });

  beforeEach(() => {
    apiStatus = 200;
    offer = {
      id: "123",
      offer_name: "Jakaś nazwa oferty",
      company_name: "Jakaś nazwa firmy",
      company_address: {
        street: "def",
        street_number: "1",
        city: "abc",
        postal_code: "00-000",
      },
      voivodeship: "Jakieś województwo",
      expiration_date: "2020-12-12",
      description:
        "Jakiś baaaaaaaaaaaaaaaaaardzo dłuuuuuuuuuuuuuuugi opis oferty pracy\n123 asdasd",
    };
    user = {
      type: userTypes.STANDARD,
      token: "123",
      data: {
        status: userStatuses.VERIFIED,
      },
    };
    jest.clearAllMocks();
  });

  it("should render without crashing", async () => {
    const { container, getByText } = render(
      <UserContext.Provider value={user}>
        <MemoryRouter>
          <JobOfferDetails match={match} />
        </MemoryRouter>
      </UserContext.Provider>
    );

    await waitForElement(() => getByText("Jakaś nazwa oferty"));

    expect(container).toMatchSnapshot();
  });

  it("should render loading alert when component is waiting for api response", async () => {
    const { getByText, queryByText } = render(
      <MemoryRouter>
        <JobOfferDetails match={match} />
      </MemoryRouter>
    );

    expect(getByText("Ładowanie", { exact: false })).toBeInTheDocument();
    expect(queryByText("Jakaś nazwa oferty")).not.toBeInTheDocument();
    await waitForElement(() => getByText("Jakaś nazwa oferty"));
  });

  it("should render error alert when api returns error", async () => {
    apiStatus = 500;
    const { getByText, queryByText } = render(
      <MemoryRouter>
        <JobOfferDetails match={match} />
      </MemoryRouter>
    );

    await waitForElement(() => getByText("Wystąpił błąd", { exact: false }));
    expect(getByText("Wystąpił błąd", { exact: false })).toBeInTheDocument();
    expect(queryByText("Jakaś nazwa oferty")).not.toBeInTheDocument();
  });

  it("should render RemoveOffer component when staff user group is jobs", async () => {
    user.type = userTypes.STAFF;
    user.data.group_type = [staffTypes.JOBS];

    const { getByText, queryByText } = render(
      <UserContext.Provider value={user}>
        <MemoryRouter>
          <JobOfferDetails match={match} />
        </MemoryRouter>
      </UserContext.Provider>
    );

    await waitForElement(() => getByText("Jakaś nazwa oferty"));
    expect(queryByText("AddCvForm")).not.toBeInTheDocument();
    expect(getByText("RemoveOffer")).toBeInTheDocument();
  });

  it("should render AddCvForm component when standard user is verified", async () => {
    const { getByText, queryByText } = render(
      <UserContext.Provider value={user}>
        <MemoryRouter>
          <JobOfferDetails match={match} />
        </MemoryRouter>
      </UserContext.Provider>
    );

    await waitForElement(() => getByText("Jakaś nazwa oferty"));
    expect(queryByText("RemoveOffer")).not.toBeInTheDocument();
    expect(getByText("AddCvForm")).toBeInTheDocument();
  });

  it("should not render AddCvForm and RemoveOffer component when user is not verified and is not a jobs staff", async () => {
    user.data.status = userStatuses.AWAITING;
    const { getByText, queryByText } = render(
      <UserContext.Provider value={user}>
        <MemoryRouter>
          <JobOfferDetails match={match} />
        </MemoryRouter>
      </UserContext.Provider>
    );

    await waitForElement(() => getByText("Jakaś nazwa oferty"));
    expect(queryByText("RemoveOffer")).not.toBeInTheDocument();
    expect(queryByText("AddCvForm")).not.toBeInTheDocument();
  });
});
