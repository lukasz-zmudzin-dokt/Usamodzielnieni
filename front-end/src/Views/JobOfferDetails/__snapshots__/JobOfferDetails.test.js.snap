// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobOfferDetails should render without crashing 1`] = `
<div>
  <div
    class="jobOfferDetails container"
  >
    <div
      class="card"
    >
      <h2
        class="card-header"
      >
        Szczegóły oferty pracy
      </h2>
      <div
        class="card-body"
      >
        <div>
          <h3>
            Jakaś nazwa oferty
          </h3>
          <div
            class="row"
          >
            <div
              class="detailsItem col-xl-4 col-md-6"
            >
              <div
                class="detailsItem__label"
              >
                Nazwa firmy
              </div>
              <div
                class="detailsItem__value"
              >
                Jakaś nazwa firmy
              </div>
            </div>
            <div
              class="detailsItem col-xl-4 col-md-6"
            >
              <div
                class="detailsItem__label"
              >
                Adres firmy
              </div>
              <div
                class="detailsItem__value"
              >
                Jakiś adres
              </div>
            </div>
            <div
              class="detailsItem col-xl-4 col-md-6"
            >
              <div
                class="detailsItem__label"
              >
                Lokalizacja
              </div>
              <div
                class="detailsItem__value"
              >
                Jakieś województwo
              </div>
            </div>
            <div
              class="detailsItem col-xl-4 col-md-6"
            >
              <div
                class="detailsItem__label"
              >
                Ważne do
              </div>
              <div
                class="detailsItem__value"
              >
                2020-12-12
              </div>
            </div>
            <div
              class="detailsItem col-xl-4 col-md-6"
            >
              <div
                class="detailsItem__label"
              >
                Branża
              </div>
              <div
                class="detailsItem__value"
              />
            </div>
            <div
              class="detailsItem col-xl-4 col-md-6"
            >
              <div
                class="detailsItem__label"
              >
                Typ
              </div>
              <div
                class="detailsItem__value"
              />
            </div>
          </div>
          <p>
            Jakiś baaaaaaaaaaaaaaaaaardzo dłuuuuuuuuuuuuuuugi opis oferty pracy
123 asdasd
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
